Change how the GEOPM Service control lock / save / restore logic works

cmcantalupo opened the pull request:
- Relates to #2062
- Fixes #2092

cmcantalupo committed:
Add test_su_restart.sh

Signed-off-by: Christopher M. Cantalupo <christopher.m.cantalupo@intel.com>

cmcantalupo committed:
Add WriteLock class to varrun module

- Make SAVE_FILES creation and deletion atomic
- Change PlatformService object to use WriteLock
- Remove read/write mode from ActiveSessions data
- Separate write lock/save/restore from session files entirely
- Fixup TestActiveSessions to reflect removal of methods
  is_write_client() and set_write_client()
- The json no longer contains the "mode" field
- Refactor file checks into separate functions
- New functions allow reuse for the WriteLock
- If control lock file exists, verify security

Signed-off-by: Christopher M. Cantalupo <christopher.m.cantalupo@intel.com>

cmcantalupo commented on service/geopmdpy/varrun.py:
```diff
@@ -571,11 +533,110 @@ def _load_session_file(self, sess_path):
+        self._VAR_PATH = var_path
+        self._LOCK_PATH = os.path.join(self._VAR_PATH, "CONTROL_LOCK")
+        # Create an empty lock file if it does not exist
+        lock_pid = self.try_lock()
```
assumes that /var/run/geopm-service exists, probably should call the secure_make_dirs() method here.

cmcantalupo commented on service/integration/test/test_su_restart.sh:
```diff
@@ -0,0 +1,68 @@
+sudo systemctl stop geopm
+sleep 1
+sudo systemctl start geopm
+wait
```
Need to check on the return value from the background process

cmcantalupo commented on service/geopmdpy/varrun.py:
```diff
@@ -179,64 +184,58 @@ def secure_read_file(path):
+
 
-    # If the path exists
+def is_secure_path(path):
```
missing doc strings

cmcantalupo commented on service/geopmdpy/varrun.py:
```diff
@@ -179,64 +184,58 @@ def secure_read_file(path):
-    return contents
+
+def is_secure_file(path, fid):
+    result = False
```
missing doc strings

cmcantalupo commented on service/geopmdpy/varrun.py:
```diff
@@ -700,11 +661,98 @@ def _load_session_file(self, sess_path):
+
+class WriteLock(object):
+    def __init__(self, var_path=GEOPM_SERVICE_VAR_PATH):
+        self._VAR_PATH = var_path
```
Add doc strings

bgeltz reviewed (changes requested):
Why is all the man page work in this PR?  It's unrelated to the write lock.

bgeltz commented on service/geopmdpy/varrun.py:
```diff
@@ -700,11 +661,98 @@ def _load_session_file(self, sess_path):
+        old_mask = os.umask(0o077)
+        try:
+            trial_count = 0
+            while self._fid is None and trial_count < 2:
```
Why try twice?  Why not more/less?
Is a sleep necessary between trials?

cmcantalupo replied:
We will try twice, the first time may be foiled by a bad file, if the second attempt is also foiled, then we want to give up.  We want to try more than once to avoid a bad file causing the lock operation to fail.

We don't need to sleep because the is_secure_file() call will rename it to an invalid name.

bgeltz commented on service/geopmdpy/varrun.py:
```diff
@@ -700,11 +661,98 @@ def _load_session_file(self, sess_path):
+
+        """
+        file_pid = None
+        self._fid.seek(0, os.SEEK_END)
```
Why seek() instead of readline()?  It might simplify this logic.

cmcantalupo replied:
Updated implementation, should be a bit easier to read.  Thanks

cmcantalupo committed:
Simplify lock file interactions to use seek less

Signed-off-by: Christopher M. Cantalupo <christopher.m.cantalupo@intel.com>

cmcantalupo committed:
Add doc strings for is_secure*() methods

Signed-off-by: Christopher M. Cantalupo <christopher.m.cantalupo@intel.com>

cmcantalupo committed:
Add extra check to TestWriteLock

Signed-off-by: Christopher M. Cantalupo <christopher.m.cantalupo@intel.com>

cmcantalupo committed:
Add doc strings and a couple helper methods to WriteLock

Signed-off-by: Christopher M. Cantalupo <christopher.m.cantalupo@intel.com>

cmcantalupo committed:
Check return code of the write session in test_su_restart.sh

Signed-off-by: Christopher M. Cantalupo <christopher.m.cantalupo@intel.com>

cmcantalupo committed:
Fix unit test to reflect change

- Order of the checks was reversed so that the directory
  warning was not dead code (directories are not regular files)

Signed-off-by: Christopher M. Cantalupo <christopher.m.cantalupo@intel.com>

cmcantalupo committed:
Add a class scoped lock member for WriteLock

- The file lock will only cover cases where more than one
  process is modifying the lock file.
- Currently only geopmd will be using the WriteLock class
  and there is expected to be only one geopmd process.
- It is not known what the mechanism is for RPC calls by dasbus
  and what parallelism if any is used to handle concurrent
  calls to the DBus interface.
- For this reason we add a member which will cover any cases
  where two concurrent calls are handled by one process.
- We will also use the advisory file lock to cover any cases where
  two calls are handled by two different processes.

Signed-off-by: Christopher M. Cantalupo <christopher.m.cantalupo@intel.com>

cmcantalupo committed:
Add bad file tests

Signed-off-by: Christopher M. Cantalupo <christopher.m.cantalupo@intel.com>

cmcantalupo committed:
Update Makefile for tests

Signed-off-by: Christopher M. Cantalupo <christopher.m.cantalupo@intel.com>

cmcantalupo committed:
Fix unbound variable error

- Error occurs as below when session PID is not valid after
  starting.

    File "/usr/lib/python3.6/site-packages/geopmdpy/varrun.py", line 698, in _load_session_file
        os.rename(sess_path, renamed_path)
    UnboundLocalError: local variable renamed_path referenced before assignment

Signed-off-by: Christopher M. Cantalupo <christopher.m.cantalupo@intel.com>

bgeltz commented on service/geopmdpy/varrun.py:
```diff
@@ -700,11 +692,185 @@ def _load_session_file(self, sess_path):
+        if self._fid == None:
+            self._rpc_unlock()
+            raise RuntimeError('Unable to open write lock securely after two tries')
+        fcntl.lockf(self._fid, fcntl.LOCK_EX)
```
If this is only remaining to future proof the service should it need to be multi-process, a comment explaining as much would be very beneficial to all our memories. :)

cmcantalupo committed:
Convert check_write_session.sh to use geopmread/geopmwrite

Signed-off-by: Christopher M. Cantalupo <christopher.m.cantalupo@intel.com>

Krebrov001 commented on service/geopmdpy_test/TestWriteLock.py:
```diff
@@ -0,0 +1,136 @@
+                with WriteLock(self._sess_path) as other_lock:
+                    self.fail('Able to create nested WriteLock context')
+            except RuntimeError as ex:
+                self.assertEqual('Attempt to modify control lock file concurrently', str(ex))
```
```python
self.assertEqual('Attempt to modify control lock file in the same process', str(ex))
```

cmcantalupo committed:
Fix issues with test_su_restart.sh

- Force the write to go through the service in case msr-safe
  is available.
- Avoid timing issue with sleep prior to stopping the service
  being too short.

Signed-off-by: Christopher M. Cantalupo <christopher.m.cantalupo@intel.com>

cmcantalupo committed:
Address review feedback and fix ci

Signed-off-by: Christopher M. Cantalupo <christopher.m.cantalupo@intel.com>

cmcantalupo committed:
Write better error message

Signed-off-by: Christopher M. Cantalupo <christopher.m.cantalupo@intel.com>
